llvm/lib/ProfileData/InstrProfWriter.cpp

```diff
@@ -1035,16 +1035,13 @@ static const char *ValueProfKindStr[] = {
 
 Error InstrProfWriter::validateRecord(const InstrProfRecord &Func) {
   for (uint32_t VK = 0; VK <= IPVK_Last; VK++) {
-    uint32_t NS = Func.getNumValueSites(VK);
-    if (!NS)
+    if (VK == IPVK_IndirectCallTarget || VK == IPVK_VTableTarget)
       continue;
+    uint32_t NS = Func.getNumValueSites(VK);
     for (uint32_t S = 0; S < NS; S++) {
-      uint32_t ND = Func.getNumValueDataForSite(VK, S);
-      std::unique_ptr<InstrProfValueData[]> VD = Func.getValueForSite(VK, S);
       DenseSet<uint64_t> SeenValues;
-      for (uint32_t I = 0; I < ND; I++)
-        if ((VK != IPVK_IndirectCallTarget && VK != IPVK_VTableTarget) &&
```

Comment:
https://github.com/llvm/llvm-project/commit/a4b543a5a541eeff5a7ba92ada4b0d809a2c8482 reports errors on duplicated values.

I'm going to be away from keyboard for sometime but curious to dig the rationale before that commit (no need to wait for my dig for this change).

Reply:
No producer can generate dups for value profile data, so there is probably no need to specialize these two kinds. Can be cleaned up later.

Reply:
Before https://github.com/llvm/llvm-project/commit/a4b543a5a541eeff5a7ba92ada4b0d809a2c8482, 'validateRecord' allows 'Value' to be zero.

I read the commit message of https://github.com/llvm/llvm-project/commit/a4b543a5a541eeff5a7ba92ada4b0d809a2c8482 a second time, and think duplicated zeros can come from valid profiles; IndirectCallTarget and VTableTarget value profiles map runtime address to functions' and global variables' GUID. If the function or global variable isn't instrumented (e.g., per IR module opt-out of `-fprofile-generate`),  (unique) runtime address might be mapped to (duplicated) zero (by `InstrProfRecord::remapValue`).

Reply:
I happen to have a raw instrumentation profile lying around for clang (that is, clang as an application).  When I generate an indexed profile with `validateRecord` modified to check all value profile kinds, including `IPVK_IndirectCallTarget` and `IPVK_VTableTarget`, then I get an `invalid_prof` error.  So, duplicate values for the two kinds seem to be a real thing.  I have yet to investigate where they come from though.

Reply:
For `IPVK_VTableTarget`, a heuristic is used to find vtables in instrumented binary (https://github.com/llvm/llvm-project/blob/2d9b6a01c7a77ee76a5c279901bca1659a550499/llvm/include/llvm/Analysis/IndirectCallVisitor.h#L45-L57), so runtime address of non-vtables can be collected.
* https://gcc.godbolt.org/z/ddMfrWaYd is an example from [rfc](https://discourse.llvm.org/t/rfc-dynamic-type-profiling-and-optimizations-in-llvm/74600#pick-instrumentation-points-and-instrument-runtime-types-7).The heuristic can find the IR sequence but it's not vtable.
* `InstrProfRecord::remapValue` will map such runtime address to zero. It should be rare for symbol GUID to be zero; in the rare case it happens, the profile use in ICP pass gates de-virtualization by comparing object pointers with vtables for correctness.